### ISSUE TITLE
Unable to convert a queryable method to an enumerable method.

### ISSUE THREAD

**Issue by jlavallet** [dotnet/efcore#26912]
I have searched both the open and closed issues. I have not found a mention of my exception.

I am creating a somewhat complex query to retrieve some many to many relationships with an entity. Here is thee code that creates the query:

```C#
      protected virtual async Task<IQueryable<ContactWithNavigationProperties>> GetQueryForNavigationPropertiesAsync()
      {
        var dbContext = await GetDbContextAsync();
        return from contact in await GetDbSetAsync()
               join department in dbContext.Departments
                 on contact.DepartmentId equals department.Id into departments
               from department in departments.DefaultIfEmpty()
               join client in dbContext.Clients
                 on contact.ClientId equals client.Id into clients
               from client in clients.DefaultIfEmpty()
               join contactAddress in dbContext.Set<ContactAddress>()
                 on contact.Id equals contactAddress.ContactId into contactAddresses
               from contactAddress in contactAddresses.DefaultIfEmpty()
               join contactEmailAddress in dbContext.Set<ContactEmailAddress>()
                 on contact.Id equals contactEmailAddress.ContactId into contactEmailAddresses
               from contactEmailAddress in contactEmailAddresses.DefaultIfEmpty()
               join contactPhoneNumber in dbContext.Set<ContactPhoneNumber>()
                 on contact.Id equals contactPhoneNumber.ContactId into contactPhoneNumbers
               from contactPhoneNumber in contactPhoneNumbers.DefaultIfEmpty()
               select new ContactWithNavigationProperties
               {
                 Contact = contact,
                 Department = department,
                 Client = client,
                 Addresses = (from ca in contactAddresses
                              join address in dbContext.Addresses
                                on ca.AddressId equals address.Id into addresses
                              from address in addresses.DefaultIfEmpty()
                              select address).ToList(),
                 EmailAddresses = (from cea in contactEmailAddresses
                                   join emailAddress in dbContext.EmailAddresses
                                     on cea.EmailAddressId equals emailAddress.Id into emailAddresses
                                   from emailAddress in emailAddresses.DefaultIfEmpty()
                                   select emailAddress).ToList(),
                 PhoneNumbers = (from cpn in contactPhoneNumbers
                                 join phoneNumber in dbContext.PhoneNumbers
                                   on cpn.PhoneNumberId equals phoneNumber.Id into phoneNumbers
                                 from phoneNumber in phoneNumbers.DefaultIfEmpty()
                                 select phoneNumber).ToList()
               };
      }
```

As soon as this query is executed, I am getting the following exception:

>    2021-12-06 07:49:17.867 -06:00 [ERR] Unable to convert a queryable method to an enumerable method. This is likely an issue in Entity Framework, please file an issue at https://go.microsoft.com/fwlink/?linkid=2142044.
    System.InvalidOperationException: Unable to convert a queryable method to an enumerable method. This is likely an issue in Entity Framework, please file an issue at https://go.microsoft.com/fwlink/?linkid=2142044.
       at Microsoft.EntityFrameworkCore.Query.Internal.NavigationExpandingExpressionVisitor.ConvertToEnumerable(MethodInfo queryableMethod, IEnumerable\`1 arguments)
       at Microsoft.EntityFrameworkCore.Query.Internal.NavigationExpandingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
       at Microsoft.EntityFrameworkCore.Query.Internal.NavigationExpandingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
       at System.Linq.Expressions.ExpressionVisitor.VisitMemberAssignment(MemberAssignment node)
       at System.Linq.Expressions.ExpressionVisitor.VisitMemberBinding(MemberBinding node)
       at System.Linq.Expressions.ExpressionVisitor.Visit[T](ReadOnlyCollection\`1 nodes, Func\`2 elementVisitor)
       at System.Linq.Expressions.ExpressionVisitor.VisitMemberInit(MemberInitExpression node)
       at System.Linq.Expressions.MemberInitExpression.Accept(ExpressionVisitor visitor)
       at Microsoft.EntityFrameworkCore.Query.Internal.NavigationExpandingExpressionVisitor.PendingSelectorExpandingExpressionVisitor.Visit(Expression expression)
       at Microsoft.EntityFrameworkCore.Query.Internal.NavigationExpandingExpressionVisitor.Expand(Expression query)
       at Microsoft.EntityFrameworkCore.Query.QueryTranslationPreprocessor.Process(Expression query)
       at Microsoft.EntityFrameworkCore.Query.RelationalQueryTranslationPreprocessor.Process(Expression query)
       at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutor[TResult](Expression query)
       at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](Expression query, Boolean async)
       at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](IDatabase database, Expression query, IModel model, Boolean async)
       at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass12_0`1.<ExecuteAsync>b__0()
       at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
       at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.ExecuteAsync[TResult](Expression query, CancellationToken cancellationToken)
       at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.ExecuteAsync[TResult](Expression expression, CancellationToken cancellationToken)
       at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryable`1.GetAsyncEnumerator(CancellationToken cancellationToken)
       at System.Runtime.CompilerServices.ConfiguredCancelableAsyncEnumerable`1.GetAsyncEnumerator()
       at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ToListAsync[TSource](IQueryable`1 source, CancellationToken cancellationToken)
       at MyProject.Contacts.EfCoreContactRepository.GetListWithNavigationPropertiesAsync(String filterText, String name, String surname, String title, Nullable`1 lastContactDateMin, Nullable`1 lastContactDateMax, String note, Nullable`1 departmentId, Nullable`1 clientId, Nullable`1 primary, Nullable`1 active, String sorting, Int32 maxResultCount, Int32 skipCount, CancellationToken cancellationToken) in D:\Century\Internal\Clients.Link\Clients.Link\src\MyProject.EntityFrameworkCore\Contacts\EfCoreContactRepository.cs:line 54
       at Castle.DynamicProxy.AsyncInterceptorBase.ProceedAsynchronous[TResult](IInvocation invocation, IInvocationProceedInfo proceedInfo)
       at Volo.Abp.Castle.DynamicProxy.CastleAbpMethodInvocationAdapterWithReturnValue`1.ProceedAsync()
       at Volo.Abp.Uow.UnitOfWorkInterceptor.InterceptAsync(IAbpMethodInvocation invocation)
       at Volo.Abp.Castle.DynamicProxy.CastleAsyncAbpInterceptorAdapter`1.InterceptAsync[TResult](IInvocation invocation, IInvocationProceedInfo proceedInfo, Func`3 proceed)
       at MyProject.Contacts.ContactsAppService.GetListAsync(GetContactsInput input) in D:\Century\Internal\Clients.Link\Clients.Link\src\MyProject.Application\Contacts\ContactAppService.cs:line 59
       at Castle.DynamicProxy.AsyncInterceptorBase.ProceedAsynchronous[TResult](IInvocation invocation, IInvocationProceedInfo proceedInfo)
       at Volo.Abp.Castle.DynamicProxy.CastleAbpMethodInvocationAdapterWithReturnValue`1.ProceedAsync()
       at Volo.Abp.Authorization.AuthorizationInterceptor.InterceptAsync(IAbpMethodInvocation invocation)
       at Volo.Abp.Castle.DynamicProxy.CastleAsyncAbpInterceptorAdapter`1.InterceptAsync[TResult](IInvocation invocation, IInvocationProceedInfo proceedInfo, Func`3 proceed)
       at Castle.DynamicProxy.AsyncInterceptorBase.ProceedAsynchronous[TResult](IInvocation invocation, IInvocationProceedInfo proceedInfo)
       at Volo.Abp.Castle.DynamicProxy.CastleAbpMethodInvocationAdapterWithReturnValue`1.ProceedAsync()
       at Volo.Abp.GlobalFeatures.GlobalFeatureInterceptor.InterceptAsync(IAbpMethodInvocation invocation)
       at Volo.Abp.Castle.DynamicProxy.CastleAsyncAbpInterceptorAdapter`1.InterceptAsync[TResult](IInvocation invocation, IInvocationProceedInfo proceedInfo, Func`3 proceed)
       at Castle.DynamicProxy.AsyncInterceptorBase.ProceedAsynchronous[TResult](IInvocation invocation, IInvocationProceedInfo proceedInfo)
       at Volo.Abp.Castle.DynamicProxy.CastleAbpMethodInvocationAdapterWithReturnValue`1.ProceedAsync()
       at Volo.Abp.Auditing.AuditingInterceptor.ProceedByLoggingAsync(IAbpMethodInvocation invocation, IAuditingHelper auditingHelper, IAuditLogScope auditLogScope)
       at Volo.Abp.Auditing.AuditingInterceptor.InterceptAsync(IAbpMethodInvocation invocation)
       at Volo.Abp.Castle.DynamicProxy.CastleAsyncAbpInterceptorAdapter`1.InterceptAsync[TResult](IInvocation invocation, IInvocationProceedInfo proceedInfo, Func`3 proceed)
       at Castle.DynamicProxy.AsyncInterceptorBase.ProceedAsynchronous[TResult](IInvocation invocation, IInvocationProceedInfo proceedInfo)
       at Volo.Abp.Castle.DynamicProxy.CastleAbpMethodInvocationAdapterWithReturnValue`1.ProceedAsync()
       at Volo.Abp.Validation.ValidationInterceptor.InterceptAsync(IAbpMethodInvocation invocation)
       at Volo.Abp.Castle.DynamicProxy.CastleAsyncAbpInterceptorAdapter`1.InterceptAsync[TResult](IInvocation invocation, IInvocationProceedInfo proceedInfo, Func`3 proceed)
       at Castle.DynamicProxy.AsyncInterceptorBase.ProceedAsynchronous[TResult](IInvocation invocation, IInvocationProceedInfo proceedInfo)
       at Volo.Abp.Castle.DynamicProxy.CastleAbpMethodInvocationAdapterWithReturnValue`1.ProceedAsync()
       at Volo.Abp.Uow.UnitOfWorkInterceptor.InterceptAsync(IAbpMethodInvocation invocation)
       at Volo.Abp.Castle.DynamicProxy.CastleAsyncAbpInterceptorAdapter`1.InterceptAsync[TResult](IInvocation invocation, IInvocationProceedInfo proceedInfo, Func`3 proceed)
       at lambda_method3642(Closure , Object )
       at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.AwaitableObjectResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
       at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)
       at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
       at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
       at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
       at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
       at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextExceptionFilterAsync>g__Awaited|25_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)

Here is my csproj file with its dependencies:

```xml
    <Project Sdk="Microsoft.NET.Sdk">
    
      <Import Project="..\..\common.props" />
    
      <PropertyGroup>
        <TargetFramework>net5.0</TargetFramework>
        <RootNamespace>MyProject</RootNamespace>
      </PropertyGroup>
    
      <ItemGroup>
        <Compile Remove="Addresses\AddressesDataSeedContributorWithDependencies.cs" />
        <Compile Remove="Campaigns\CampaignsDataSeedContributor.cs" />
        <Compile Remove="ClientNeeds\ClientNeedsDataSeedContributor.cs" />
        <Compile Remove="Clients\ClientsDataSeedContributorWithDependencies.cs" />
        <Compile Remove="Contacts\ContactsDataSeedContributorWithDependencies.cs" />
        <Compile Remove="Countries\CountriesDataSeedContributor.cs" />
        <Compile Remove="Departments\DepartmentsDataSeedContributor.cs" />
        <Compile Remove="EmailAddresses\EmailAddressesDataSeedContributor.cs" />
        <Compile Remove="EntityFrameworkCore\MyProjectDataSeeder.cs" />
        <Compile Remove="LeadSources\LeadSourcesDataSeedContributor.cs" />
        <Compile Remove="Migrations\MyProjectDbContextModelSnapshot - Copy.cs" />
        <Compile Remove="Opportunities\OpportunitiesDataSeedContributorWithDependencies.cs" />
        <Compile Remove="Opportunities\OpportunityLineItemsDataSeedContributorWithDependencies.cs" />
        <Compile Remove="PhoneNumbers\PhoneNumbersDataSeedContributor.cs" />
        <Compile Remove="ProductGroups\ProductGroupsDataSeedContributor.cs" />
        <Compile Remove="Products\ProductsDataSeedContributorWithDependencies.cs" />
        <Compile Remove="RevenueRepeatFreqs\RevenueRepeatFreqsDataSeedContributor.cs" />
        <Compile Remove="States\StatesDataSeedContributorWithDependencies.cs" />
      </ItemGroup>
    
      <ItemGroup>
        <ProjectReference Include="..\MyProject.Domain\MyProject.Domain.csproj" />
        <PackageReference Include="Microsoft.Data.SqlClient" Version="3.0.1" />
        <PackageReference Include="Volo.Abp.EntityFrameworkCore.SqlServer" Version="4.4.4" />
        <PackageReference Include="Volo.Abp.PermissionManagement.EntityFrameworkCore" Version="4.4.4" />
        <PackageReference Include="Volo.Abp.SettingManagement.EntityFrameworkCore" Version="4.4.4" />
        <PackageReference Include="Volo.Abp.IdentityServer.EntityFrameworkCore" Version="4.4.4" />
        <PackageReference Include="Volo.Abp.BackgroundJobs.EntityFrameworkCore" Version="4.4.4" />
        <PackageReference Include="Volo.Abp.AuditLogging.EntityFrameworkCore" Version="4.4.4" />
        <PackageReference Include="Volo.Abp.FeatureManagement.EntityFrameworkCore" Version="4.4.4" />
        <PackageReference Include="Volo.Abp.BlobStoring.Database.EntityFrameworkCore" Version="4.4.4" />
        <PackageReference Include="Volo.Abp.Identity.Pro.EntityFrameworkCore" Version="4.4.4" />
        <PackageReference Include="Volo.Abp.LanguageManagement.EntityFrameworkCore" Version="4.4.4" />
        <PackageReference Include="Volo.Saas.EntityFrameworkCore" Version="4.4.4" />
        <PackageReference Include="Volo.Abp.TextTemplateManagement.EntityFrameworkCore" Version="4.4.4" />
        <PackageReference Include="Volo.CmsKit.Pro.EntityFrameworkCore" Version="4.4.4" />
      </ItemGroup>
    
      <ItemGroup>
        <PackageReference Include="Microsoft.EntityFrameworkCore.Tools" Version="5.0.*">
          <IncludeAssets>runtime; build; native; contentfiles; analyzers</IncludeAssets>
          <PrivateAssets>compile; contentFiles; build; buildMultitargeting; buildTransitive; analyzers; native</PrivateAssets>
        </PackageReference>
      </ItemGroup>
    
    </Project>
```

As you can see I am using ABP Commercial 4.4.4 with EF 5.0. Thanks for your help.

EF Core version:  5.0.12
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET 5.0
Operating system:  Microsoft Windows 11 Pro
IDE: Visual Studio 2022 17.0.2

**Comment by ajcvickers** [comment]
@jlavallet We discussed this, and there is a reasonable chance based on the information above that this could be fixed in EF Core 6.0. Can you give that a try? If it still fail, then please attach a small, runnable project or post a small, runnable code listing that reproduces what you are seeing so that we can investigate.

**Comment by ajcvickers** [comment]
**EF Team Triage:** Closing this issue as the requested additional details have not been provided and we have been unable to reproduce it.

*BTW this is a canned response and may have info or details that do not directly apply to this particular issue. While we'd like to spend the time to uniquely address every incoming issue, we get a lot traffic on the EF projects and that is not practical. To ensure we maximize the time we have to work on fixing bugs, implementing new features, etc. we use canned responses for common triage decisions.*
